src/Bedrock.Framework.Experimental/Transports/WebSockets/WebSocketPayloadEncoder.cs

```diff
@@ -0,0 +1,152 @@
+
+        /// <summary>
+        /// Masks or unmasks a WebSocket payload span.
+        /// </summary>
+        /// <param name="span">The span to mask or unmask.</param>
+        /// <param name="bytesToRead">The number of bytes to read from the span.</param>
+        /// <param name="useSimd">Whether or not to use SIMD accelerated masking.</param>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private unsafe void MaskUnmaskSpan(in ReadOnlySpan<byte> span, long bytesToRead, bool useSimd)
+        {
+            var maskingKey = _maskingKey;
+
+            fixed (byte* dataStartPtr = &MemoryMarshal.GetReference(span))
+            {
+                byte* dataPtr = dataStartPtr;
+                byte* dataEndPtr = dataStartPtr + bytesToRead;
+                byte* maskPtr = (byte*)&maskingKey;
+
+                //If there is at least one int of data available
+                if (dataEndPtr - dataStartPtr >= sizeof(int))
+                {
+                    //Start by aligning to an int boundary, makes aligned loads faster later
+                    while ((ulong)dataPtr % sizeof(int) != 0)
+                    {
+                        Debug.Assert(dataPtr < dataEndPtr);
+
+                        *dataPtr++ ^= maskPtr[_currentMaskIndex];
+                        _currentMaskIndex = (_currentMaskIndex + 1) & 3; //Modulus shortcut for power of 2 values for perf, same as _currentMaskIndex % 4
+                    }
+
+                    //We may have moved forward an uneven number of bytes along the masking key,
+                    //generate a new masking key that has those bytes rotated for whole-int unmasking
+                    int alignedMask = (int)BitOperations.RotateRight((uint)_maskingKey, _currentMaskIndex * 8);
+
+                    //If we found we should use SIMD and there is sufficient data to read
+                    if (useSimd && (dataEndPtr - dataPtr) >= Vector<byte>.Count)
+                    {
+                        //Align by whole ints to full SIMD load boundary to avoid a perf penalty for unaligned loads
+                        while ((ulong)dataPtr % (uint)Vector<byte>.Count != 0)
+                        {
+                            Debug.Assert(dataPtr < dataEndPtr);
+
+                            *(int*)dataPtr ^= alignedMask;
+                            dataPtr += sizeof(int);
+                        }
+
+                        //Unmask full aligned vectors at a time
+                        if (dataEndPtr - dataPtr >= Vector<byte>.Count)
+                        {
+                            Vector<byte> maskVector = Vector.AsVectorByte(new Vector<int>(alignedMask));
+
+                            do
+                            {
+                                *(Vector<byte>*)dataPtr ^= maskVector;
+                                dataPtr += Vector<byte>.Count;
+                            }
+                            while (dataEndPtr - dataPtr >= Vector<byte>.Count);
```

Comment:
This will result in a subtraction in each iteration.
```suggestion
                            while (dataPtr <= dataEndPtr - Vector<byte>.Count);
```
will avoid this and keeps the "constant" right side in a register -- potentially, but I'm pretty sure, especially if the `if` above is written with the same condition.

Reply:
I suppose I could calc `dataEndPtr - Vector<byte>.Count` as a variable, since that will never change, and give the JIT a leg up on registerifying it.

Reply:
I believe / expect the JIT will do this for you (as I've seen in some of my PRs which have similar checks). Especially if the `if` and the loop condition are written similar.